src/Money.php

```diff
@@ -227,158 +208,122 @@ public function add(Money ...$addends)
      */
-    public function mod(Money $divisor)
+    public function mod(Money $divisor): Money
     {
-        $this->assertSameCurrency($divisor);
+        // Note: non-strict equality is intentional here, since `Currency` is `final` and reliable.
+        if ($this->currency != $divisor->currency) {
+            throw new InvalidArgumentException('Currencies must be identical');
+        }
 
-        return new self($this->getCalculator()->mod($this->amount, $divisor->amount), $this->currency);
+        return new self(self::$calculator::mod($this->amount, $divisor->amount), $this->currency);
     }
 
     /**
      * Allocate the money according to a list of ratios.
      *
+     * @psalm-param TRatios $ratios
+     *
      * @return Money[]
+     * @psalm-return (
+     *     TRatios is list
+     *         ? non-empty-list<Money>
+     *         : non-empty-array<Money>
+     * )
+     *
+     * @template TRatios as non-empty-array<float|int>
      */
-    public function allocate(array $ratios)
+    public function allocate(array $ratios): array
     {
-        if (count($ratios) === 0) {
-            throw new \InvalidArgumentException('Cannot allocate to none, ratios cannot be an empty array');
-        }
-
         $remainder = $this->amount;
-        $results = [];
-        $total = array_sum($ratios);
+        $results   = [];
+        $total     = array_sum($ratios);
 
         if ($total <= 0) {
-            throw new \InvalidArgumentException('Cannot allocate to none, sum of ratios must be greater than zero');
+            throw new InvalidArgumentException('Cannot allocate to none, sum of ratios must be greater than zero');
         }
 
         foreach ($ratios as $key => $ratio) {
             if ($ratio < 0) {
-                throw new \InvalidArgumentException('Cannot allocate to none, ratio must be zero or positive');
+                throw new InvalidArgumentException('Cannot allocate to none, ratio must be zero or positive');
             }
-            $share = $this->getCalculator()->share($this->amount, $ratio, $total);
-            $results[$key] = $this->newInstance($share);
-            $remainder = $this->getCalculator()->subtract($remainder, $share);
+
+            $share         = self::$calculator::share($this->amount, (string) $ratio, (string) $total);
+            $results[$key] = new self($share, $this->currency);
+            $remainder     = self::$calculator::subtract($remainder, $share);
         }
 
-        if ($this->getCalculator()->compare($remainder, '0') === 0) {
+        if (self::$calculator::compare($remainder, '0') === 0) {
             return $results;
         }
 
-        $fractions = array_map(function ($ratio) use ($total) {
-            $share = ($ratio / $total) * $this->amount;
+        $amount    = $this->amount;
+        $fractions = array_map(static function (float|int $ratio) use ($total, $amount) {
+            $share = (float) $ratio / $total * (float) $amount;
```

Comment:
I did not really understand all of this logic, but left it as-is.

Seems risky due to divisions/multiplications happening directly here, rather than in the calculators.

Trying to replace it with calculators leads to failed tests, hence why I held back in trying to adjust it.